### DISCUSSION TITLE
Phobos interdependency and public unit tests

### DISCUSSION THREAD

**Discussion by thewilsonator** [dlang/PhobosV3-Design#33] [General]
From this months beerconf about reducing/documenting the interdependencies between phobos modules:

Pull out publicly documented unittests and verify that no any/all selective imports are present in the unittest and no global imports from that module.

```
module std.foo;

import std.bar; // a global import

struct Baz { ... }

///
@safe unittest {
    import std.quux : something;
   assert 
}
```
->
```
import std.foo;
///
@safe unittest {
    import std.quux : something;
}
```
and check this still compiles.

**Comment by jmdavis** [comment]
The tooling isn't set up to do anything with the Phobos v3 code right now, but don't we already have a tool that's doing this for Phobos v2?

**Reply by pbackus** [reply]
Yes, it's [`tests_extractor.d`][1] in the [`dlang/tools` repository][2], and it's invoked for Phobos (v2) with `make publictests`.

[1]: https://github.com/dlang/tools/blob/master/tests_extractor.d
[2]: https://github.com/dlang/tools/

**Reply by thewilsonator** [reply]
NB the context for this was for phobosV2 for people that want to contribute that we can prepare ahead of time to disentangle, make the unit tests better for documentation.

**Reply by jmdavis** [reply]
I'm not sure that it would be easy to talk Walter into it, but I wonder if we should look at adding a `pragma` that did this so that projects in general could mark their ddoc-ed `unittest` blocks with it and catch when they accidentally use a  private symbol or failed to have all of the necessary imports within an example.
